smart_open/gcs.py

```diff
@@ -470,52 +479,136 @@ def terminate(self):
     #
     def _upload_next_part(self):
         part_num = self._total_parts + 1
-        logger.info(
-            "uploading part #%i, %i bytes (total %.3fGB)",
-            part_num,
-            self._current_part.tell(),
-            self._total_size / 1024.0 ** 3
-        )
-        content_length = end = self._current_part.tell()
-        start = self._total_size - content_length
-        stop = self._total_size - 1
+
+        # upload the largest amount possible given GCS's restriction
+        # of parts being multiples of 256kB, except for the last one
+        size_of_leftovers = self._current_part_size % self._min_part_size
+        content_length = self._current_part_size - size_of_leftovers
+
+        # a final upload of 0 bytes does not work, so we need to guard against this edge case
+        # this results in occasionally keeping an additional 256kB in the buffer after uploading a part,
+        # but until this is fixed on Google's end there is no other option
+        # https://stackoverflow.com/questions/60230631/upload-zero-size-final-part-to-google-cloud-storage-resumable-upload
+        if size_of_leftovers == 0:
+            content_length -= _REQUIRED_CHUNK_MULTIPLE
+
+        total_size = self._bytes_uploaded + content_length
+
+        start = self._bytes_uploaded
+        stop = total_size - 1
 
         self._current_part.seek(0)
 
         headers = {
             'Content-Length': str(content_length),
-            'Content-Range': _make_range_string(start, stop, end)
+            'Content-Range': _make_range_string(start, stop, _UNKNOWN_FILE_SIZE),
         }
-        response = self._session.put(self._resumable_upload_url, data=self._current_part, headers=headers)
 
-        if response.status_code not in _SUCCESSFUL_STATUS_CODES:
+        logger.info(
+          "uploading part #%i, "
+          "%i bytes (total %.3fGB)"
+          "headers %r",
+          part_num,
+          content_length,
+          total_size / 1024.0 ** 3,
+          headers,
+        )
+
+        response = self._session.put(
+            self._resumable_upload_url,
+            data=self._current_part.read(content_length),
+            headers=headers,
+        )
+
+        if response.status_code != _UPLOAD_INCOMPLETE_STATUS_CODE:
             msg = (
                 "upload failed ("
-                "status code: %i"
-                "response text=%s, "
+                "status code: %i "
```

Comment:
This formatting looks a little awkward, and you perform it in multiple places. Do:

```python
raise UploadFailedError.from_response(response, ...)
```

and implement the above as a `classmethod`.